controllers/kyma_controller.go

```diff
@@ -90,95 +91,95 @@ type KymaReconciler struct {
 	logger := ctrlLog.FromContext(ctx)
 	logger.V(log.InfoLevel).Info("reconciling")
 
 	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
-	// check if kyma resource exists
+
 	kyma := &v1beta2.Kyma{}
 	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
-		// we'll ignore not-found errors, since they can't be fixed by an immediate
-		// requeue (we'll need to wait for a new notification), and we can get them
-		// on deleted requests.
 		if apierrors.IsNotFound(err) {
-			// TODO: revisit this after runtime-controller get upgraded to newer version
+			// TODO: revisit after runtime-controller gets upgraded
 			// Related issue: https://github.com/kyma-project/lifecycle-manager/issues/579
-			logger.V(log.DebugLevel).Info(fmt.Sprintf("can not found Kyma %s, assume deleted", req.NamespacedName))
+			logger.V(log.DebugLevel).Info(fmt.Sprintf("Kyma %s not found, probably already deleted", req.NamespacedName))
 		}
 
-		return ctrl.Result{}, client.IgnoreNotFound(err) //nolint:wrapcheck
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	status.InitConditions(kyma, kyma.SyncEnabled(), r.WatcherEnabled(kyma))
+	status.InitConditions(kyma, r.WatcherEnabled(kyma))
 
 	if kyma.SkipReconciliation() {
 		logger.V(log.DebugLevel).Info("kyma gets skipped because of label")
 		return ctrl.Result{RequeueAfter: r.RequeueIntervals.Success}, nil
 	}
 
+	return r.reconcile(ctx, kyma)
+}
+
+func (r *KymaReconciler) reconcile(ctx context.Context, kyma *v1beta2.Kyma) (ctrl.Result, error) {
 	if kyma.SyncEnabled() {
 		var err error
-		if ctx, err = remote.InitializeSyncContext(ctx, kyma, r.RemoteSyncNamespace,
-			remote.NewClientWithConfig(r.Client, r.KcpRestConfig), r.RemoteClientCache); err != nil {
-			err := fmt.Errorf("initializing sync context failed: %w", err)
-			r.Event(kyma, "Warning", string(SyncContextError), err.Error())
-			return r.CtrlErr(ctx, kyma, err)
+		remoteClient := remote.NewClientWithConfig(r.Client, r.KcpRestConfig)
+		if ctx, err = remote.InitializeSyncContext(ctx, kyma,
+			r.RemoteSyncNamespace, remoteClient, r.RemoteClientCache); err != nil {
+			r.enqueueWarningEvent(kyma, syncContextError, err)
+			return r.requeueWithError(ctx, kyma, err)
 		}
 	}
 
-	// check if deletionTimestamp is set, retry until it gets fully deleted
 	if !kyma.DeletionTimestamp.IsZero() && kyma.Status.State != v1beta2.StateDeleting {
-		return r.deleteKyma(ctx, kyma)
+		if err := r.deleteRemoteKyma(ctx, kyma); err != nil {
+			return r.requeueWithError(ctx, kyma, err)
+		}
+		if err := r.updateStatus(ctx, kyma, v1beta2.StateDeleting, "waiting for modules to be deleted"); err != nil {
+			return r.requeueWithError(ctx, kyma, fmt.Errorf("could not update kyma status after triggering deletion: %w", err))
+		}
+		return ctrl.Result{}, nil
 	}
 
-	// check finalizer
-	if kyma.CheckLabelsAndFinalizers() {
+	if needsUpdate := kyma.EnsureLabelsAndFinalizers(); needsUpdate {
```

Comment:
what this reason to have this `needsUpdate`?

Reply:
Readability in terms of why this method returns a bool.